dask/dataframe/io/csv.py

```diff
@@ -215,21 +220,32 @@ def text_blocks_to_pandas(reader, block_lists, header, head, kwargs,
     columns = list(head.columns)
     delayed_pandas_read_text = delayed(pandas_read_text, pure=True)
     dfs = []
+
     for blocks in block_lists:
         if not blocks:
             continue
+        if include_path_column:
+            path, blocks = blocks
+            path = (include_path_column, path)
+        else:
+            path = None
         df = delayed_pandas_read_text(reader, blocks[0], header, kwargs,
                                       dtypes, columns, write_header=False,
-                                      enforce=enforce)
+                                      enforce=enforce, path=path)
+
         dfs.append(df)
         rest_kwargs = kwargs.copy()
         rest_kwargs.pop('skiprows', None)
         for b in blocks[1:]:
             dfs.append(delayed_pandas_read_text(reader, b, header, rest_kwargs,
                                                 dtypes, columns,
-                                                enforce=enforce))
+                                                enforce=enforce, path=path))
 
     if collection:
+        if include_path_column:
+            head = head.assign(**{include_path_column: 'path_to_file'})
+            head[include_path_column] = head[include_path_column].astype('category')
+            unknown_categoricals = head.select_dtypes(include=['category']).columns
```

Comment:
The categories *are* known though, so you'll want to do something like (untested):

```
# Assumes you have a list of the paths, as returned from `read_bytes`
head = head.assign(**{include_path_column: pd.Categorical.from_codes(range(len(paths)), paths)})
```

More information on how dask handles categories (known/unknown) is here: http://dask.pydata.org/en/latest/dataframe-design.html#categoricals

Reply:
hmm. Wouldn't that require that the length of `head` be at least n files?

Reply:
Oop, good catch (hence the untested :)). Should be more like:

```
head = head.assign(**{include_path_column: pd.Categorical.from_codes(np.zeros(len(head)), paths)})
```

The intent is that `head` includes the categoricals in the proper order for all files, it doesn't really matter what the value is set to for those rows. Head just includes the correct dtypes.

Reply:
Cool! So based on this discussion it seems like we want a list of paths from `read_bytes` rather than a tuple of (path, blocklist) for each file.